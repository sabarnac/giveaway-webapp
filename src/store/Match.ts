import { observable, computed, action } from "mobx";
import Participant from "./Participant";
import Random from "random-js";
import Config from "../config/Config";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _participants: Participant[];
  @observable private _winner: Participant;
  private static _RANDOM_GENERATOR: Random = Config.randomGenerator;

  private _getWinner = (): Participant =>
    Match._RANDOM_GENERATOR.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  public get winner(): Participant {
    return this._winner;
  }
}
