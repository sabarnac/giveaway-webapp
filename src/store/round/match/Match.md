Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const match: Match = new Match(
  [new Participant("FOO"), new Participant("BAR")],
  "1",
);
const matchId: string = match.id; // "1"
const matchCompleteId: string = match.fullId; // "1:1"
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
const winner: Participant[] = round.winner; // Participant("FOO")
```
