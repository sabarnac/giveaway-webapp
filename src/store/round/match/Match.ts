import { observable, computed } from "mobx";
import Participant from "./participant/Participant";
import RandomGenerator from "../../config/RandomGenerator";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _id: string;
  @observable private _participants: Participant[];
  @observable private _winner: Participant;

  private static counter: number = 1;

  private _getWinner = (): Participant =>
    RandomGenerator.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._id = `match-${Match.counter++}`;
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the match ID.
   * @returns {string} The unique ID of the match.
   */
  @computed public get id(): string {
    return this._id;
  }

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  @computed public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  @computed public get winner(): Participant {
    return this._winner;
  }

  /**
   * Determines whether another match is identical to the current one.
   * @param  {Match} otherMatch The other match to compare against.
   * @returns {boolean} Whether the other match is equal to the current one.
   */
  public equals = (otherMatch: Match): boolean => this._id === otherMatch._id;
}
