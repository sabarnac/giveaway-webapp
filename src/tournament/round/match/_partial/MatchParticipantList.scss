.match__list {
  width: 50%;
  & > .participant-entry {
    border-bottom: none;
    &:last-child {
      border-bottom: 1px solid #000000;
    }
  }
}
