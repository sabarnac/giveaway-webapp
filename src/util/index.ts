export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
