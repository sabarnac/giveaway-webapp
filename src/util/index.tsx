import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
