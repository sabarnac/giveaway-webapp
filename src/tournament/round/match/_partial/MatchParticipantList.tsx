import MatchParticipantList from "./MatchParticipantList.react";

/**
 * React component for the match participant list.
 */
export default MatchParticipantList;
