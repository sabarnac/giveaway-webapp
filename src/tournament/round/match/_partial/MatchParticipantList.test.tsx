import React from "react";
import MatchParticipantList from "./MatchParticipantList.react";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyMatch } from "../../../../util/test";
import Match from "../../../../store/round/match/Match";

describe("MatchParticipantList Component.", () => {
  it("Roundes snapshot.", () => {
    const match: Match = createDummyMatch();
    const component: ReactTestRenderer = create(
      <MatchParticipantList className="foobar" match={match} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
