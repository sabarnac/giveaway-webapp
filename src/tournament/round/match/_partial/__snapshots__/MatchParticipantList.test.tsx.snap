// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchParticipantList Component. Roundes snapshot. 1`] = `
<div
  className="foobar__list"
>
  <div
    className="participant participant-entry"
  >
    <div
      className="participant-entry__avatar"
    >
      <img
        alt="barfoo-1"
        src="foobar-1"
      />
    </div>
    <div
      className="participant-entry__name"
    >
      <strong>
        Fozbaz 1
      </strong>
    </div>
  </div>
  <div
    className="participant participant-entry"
  >
    <div
      className="participant-entry__avatar"
    >
      <img
        alt="barfoo-2"
        src="foobar-2"
      />
    </div>
    <div
      className="participant-entry__name"
    >
      <strong>
        Fozbaz 2
      </strong>
    </div>
  </div>
</div>
`;
